### ISSUE TITLE
Training Tokenizer and Vocoder

### ISSUE THREAD

**Issue by dox012** [slp-rl/slamkit#16]
Thank you for providing the training script. I successfully ran the English pretraining, and the results look quite good. Now I’d like to pretrain on Chinese data. Do I need to retrain HuBERT and HiFi-GAN for this? How can I train these two components? Can textlesslib be used to train them?

**Comment by gallilmaimon** [comment]
We appreciate your interest in our work, I'd be very interested to see it used for other languages as well :)

> Do I need to retrain HuBERT and HiFi-GAN for this?

I suspect you would as the quality of re-synthesis will probably be poor despite the encoder being "multi-lingual". The best way to check, is to encode and vocode some speech segments without sampling from the model. If the re-synthesised audio sounds poor for Chinese that is probably a top limit for how well sampling from any model would be. Another option would be to only fit the k-means and then train hifigan (using the existing HuBERT), which might provide decent results without needing to train HuBERT which is more compute intensive.

Another option would be to use other existing speech tokenisers for Chinese, but that might require writing a custom [hubert_feature_extractor.py](https://github.com/slp-rl/slamkit/blob/main/slamkit/feature_extractor/hubert_feature_extractor.py)

> How can I train these two components? Can textlesslib be used to train them?

The most direct way would be to use textlesslib to train these components, as our HuBERT and HiFiGAN are from there. The hifi-GAN should be able to be loaded directly with our code which was ported/isolated from there. 
Regarding HuBERT after training you could convert the weights from textlesslib to HF format (as we did) using the [official conversion script]( https://github.com/huggingface/transformers/blob/main/src/transformers/models/hubert/convert_hubert_original_pytorch_checkpoint_to_pytorch.py). After that, our class should be able to load it directly:
https://github.com/slp-rl/slamkit/blob/388b40222a1736930dad5e6b231754217c69629a/slamkit/feature_extractor/hubert_feature_extractor.py#L16

Note that you will need to place the k-means in the correct local directory/server.

Another alternative would be to implement a custom feature extractor which inherits from our base feature extractor and then one can use textlesslib or any other dependancies, we wanted to keep the core dependencies minimal.


Please let me know if you have any further questions:) When it comes to actually training the SpeechLM itself I suggest creating synthetic TTS training sets for pre-training and DPO as we saw these boosted performance notably. I saw that the TTS we used for DPO data creation has a Chinese version, but I can not speak to it's quality - https://huggingface.co/hexgrad/Kokoro-82M-v1.1-zh

**Comment by dox012** [comment]
I tried training a model using AISHELL (a Chinese speech dataset) with the original configuration, without modifying the tokenizer or vocoder. It seems that the generated speech sounds like Chinese in terms of content, although the voice timbre resembles a foreign language. This suggests that the model might be able to perform Chinese continuation to some extent, but it indeed requires retraining the tokenizer and vocoder. Currently, I'm working on training these two components using textlesslib. (GitHub doesn’t support .wav attachments, so I’m unable to upload the generated results.)

**Comment by gallilmaimon** [comment]
Yeah, that makes sense and does sound like a known issue with using vocoders for languages that they weren't trained for. Let us know how it works with yours, and if you come across any issues adapting the library to your use case!

FYI, you can upload zipped audio to Github, though I do not speak Chinese and therefore won't be able to assess the quality or correctness anyway 😅

**Comment by dox012** [comment]
I uploaded the audio generated from previous training using English HuBERT and HiFi-GAN on mixed English & Chinese dataset for reference. Although the timbre sounds quite similar, the continuation in Chinese has different pauses and rhythms compared to English, which suggests that pre-training on mixed-language data is at least partially successful. If the model is trained only on English data, then continuing with Chinese input doesn’t produce this kind of strange sound — instead, it switches back to normal English output.

[mix.zip](https://github.com/user-attachments/files/21097967/mix.zip)
（generate_0，1，2 is Chinese continuation，generate_3，4，5 is English continuation）

After some experimentation, I successfully retrained a multilingual codec, and the reconstruction quality sounds pretty good. I'm now trying to pre-train on the same data using this new codec. Which scripts do I need to modify to enable this?

Also, I'm curious — why are there two preprocessing scripts, `extract_features.py` and `prepare_tokens.py`? I looked through the contents of the files generated in both stages and they seem mostly the same. Another question is, is HuBERT only used for preprocessing? I saw that the training script has a HuBERT tokenizer option, but as far as I understand, UnitLM takes preprocessed tokens as input. So during training, HuBERT shouldn't be needed anymore. I'm a bit confused.

**Comment by gallilmaimon** [comment]
Is your multi-lingual codec a HuBERT + HiFigan type model trained with textlesslib? If so the two options would be to convert the hubert to a hugginface format with the script I provided above and then train it "regularly" using the exact same code and pipeline as regular pre-training. If instead you prefer to use the textlesslib implementation you would need to implement a custom feature_extractor class like suggested in my previous suggestion. If you provide more information about your use case I will be able to provide more specific insights.

Regarding two pre-processing scripts - we decouple extracting tokeniser indices (`extract_features.py`) and mapping these to exact training tokens (`prepare_tokens.py`). There are several reasons for this, but mainly that extracting Hubert indices is quite slow, and some training configurations might require slightly different mappings of these indices to tokens. This is mainly true for settings where one trains interleaved Speech-Text LMs (as we did using this codebase for https://arxiv.org/abs/2504.02398), but could also be the case if you want to use different special tokens or something like that. Anyway, the prepare tokens step should be pretty fast.

**Comment by dox012** [comment]
After some modifications, I successfully ran the Chinese pretraining. This time, the pronunciation sounds more natural, but there's still a lot of room for improvement. For example, the consistency and coherence are lacking — while it sounds like Chinese, it lacks complete sentence-level semantics. I suspect this might be because the codec carries both semantic and acoustic information compared to HuBERT. However, we still want to preserve acoustic information to support generation of different speaker timbres. How can we solve this problem?

[ch.zip](https://github.com/user-attachments/files/21207859/ch.zip)
The beginning contains some noise, which might be due to minor bugs. The middle part sounds normal.

On the other hand, we're experimenting with the effects of SFT and DPO. I only found the DPO script in the code — why are we performing DPO without doing SFT first? In text models, it's common to perform DPO after the SFT stage. Would it be feasible to follow the DPO approach, using a TTS model to synthesize SFT audio data from SFT text data?

**Comment by gallilmaimon** [comment]
**Regarding using an audio codec instead of HuBERT** - if you mean that you are using an audio tokeniser based on a reconstruction objective with RVQ or SVQ for instance, it is well documented that this does not currently work well for unconditional SpeechLM (see for instance this paper - https://arxiv.org/abs/2506.10274). While there are some methods which do build on a hybrid tokeniser (i.e. one with reconstruction + some semantic distillation objective) which are used for SpeechLM such as [Mimi](https://arxiv.org/abs/2410.00037). They were trained with a text-stream as well, as many many more hours of speech data compared to the methods described here. This richer information to model could require more data and compute to model well. Also note that our existing codebase doesn't support multiple streams in the SpeechLM, but we are happy to accept PRs:)

**Regarding DPO vs SFT** - DPO is essentially a comparative loss objective, and as such can be used for various motivations as part of the training pipeline. In the text domain it is usually used to steer strong instruction tuned LLM to better/more aligned answers. However, in this context the SpeechLM is a base model (not instruction tuned, no special user/system tags etc). Therefore, we use DPO with synthetic data (in [Slamming](https://arxiv.org/abs/2502.15814)) only to boost semantic consistency and grammar abilities etc (based on TTS of [SWAG](https://huggingface.co/datasets/allenai/swag)). We saw that in our setup this was better than pre-training/SFT on the positive examples only. The DPO trainer also have various options some of which would let you compute the loss only on the continuation and not on the prompt, which is can be used for a sort of SFT. You can definitely use the DPOTrainer also for other objective and definitely create TTS data also for pre-training or some type of SFT. I suspect that could further boost performance if done well.
